File: Problem93_CustomSortString.java
// Time Complexity: O(m + n)
// Space Complexity: O(1)

import java.util.*;
class Solution {
    public String customSortString(String order, String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        for(int i = 0; i < s.length(); i++) {
            map.put(s.charAt(i), map.getOrDefault(s.charAt(i), 0) + 1);
        }
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < order.length(); i++) {
            char c = order.charAt(i);
            if(map.containsKey(c)) {
                int count = map.get(c);
                while(count > 0) {
                    sb.append(c);
                    count--;
                }
            }
            map.remove(c);
        }
        for(char c: map.keySet()) {
            int count = map.get(c);
            while(count > 0) {
                sb.append(c);
                count--;
            }
        }
        return sb.toString();
    }
}

File: Problem-93 Custom Sort String.py
# 791. Custom Sort String
# https://leetcode.com/problems/custom-sort-string/

# Logic: Iterate over s and make a count hashmap. Iteratrate over orders and add to the 
# result number of times the letter exists in the hashmap and also remove the entry from 
# hashmap. Iterate over the remaining keys in the hashmap and add to result.

# Time Complexiety: O(n)
# Space Complexiety: O(1) [worst case it will be O(26) if all the letters are to be put in the hashmap]

class Solution:
    def customSortString(self, order: str, s: str) -> str:
        hashmap = dict()
        
        for i in s:
            if i in hashmap:
                hashmap[i] += 1
            else:
                hashmap[i] = 1
        
        res = list()
        
        for i in order:
            if i in hashmap:
                for _ in range(hashmap[i]):
                    res.append(i)
                hashmap.pop(i)
        
        for i in hashmap:
            for j in range(hashmap[i]):
                res.append(i)
        
        return "".join(res)